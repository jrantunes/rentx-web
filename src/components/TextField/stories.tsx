import { ComponentMeta, ComponentStory } from "@storybook/react";
import { AiOutlineMail } from "react-icons/ai";
import TextField from ".";

export default {
  title: "Form/TextField",
  component: TextField,
  args: {
    label: "Label",
    labelFor: "input",
    placeholder: "Placeholder",
    disabled: false,
    iconPosition: "left"
  },
  argTypes: {
    labelFor: {
      control: false
    },
    icon: {
      control: false
    },
    error: {
      control: false
    },
    initialValue: {
      control: false
    },
    iconPosition: {
      control: "select"
    }
  }
} as ComponentMeta<typeof TextField>;

export const Default: ComponentStory<typeof TextField> = (args) => (
  <div style={{ maxWidth: 340, padding: 15 }}>
    <TextField {...args} />
  </div>
);

export const withIcon: ComponentStory<typeof TextField> = (args) => (
  <div style={{ maxWidth: 340, padding: 15 }}>
    <TextField {...args} />
  </div>
);

withIcon.args = {
  icon: <AiOutlineMail />
};

export const withError: ComponentStory<typeof TextField> = (args) => (
  <div style={{ maxWidth: 340, padding: 15 }}>
    <TextField {...args} />
  </div>
);

withError.args = {
  error: "Algum erro aconteceu!",
  initialValue: "Errou!!!",
  icon: <AiOutlineMail />
};

withError.argTypes = {
  disabled: {
    control: false
  },
  error: {
    control: "string"
  }
};
