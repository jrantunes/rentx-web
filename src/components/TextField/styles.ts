import styled, { css, DefaultTheme } from "styled-components";
import { TextFieldProps } from ".";

type WrapperProps = Pick<TextFieldProps, "disabled"> & {
  error: boolean;
  hasValue: boolean;
};

type HasValueProps = {
  hasValue: boolean;
};

type IconPositionProps = {
  iconPosition: "left" | "right";
};

export const InputWrapper = styled.div`
  ${({ theme }) => css`
    display: flex;
    align-items: center;
    background-color: ${theme.colors.white};
    padding: 0 ${theme.spacings.xsmall};
    border: 0.1rem solid ${theme.colors.gray200};
    transition: border ${theme.transitions.fast};

    &:hover,
    &:focus-within {
      border-color: ${theme.colors.gray400};
    }

    cursor: text;
  `}
`;

type InputContentProps = {
  hasIcon: boolean;
} & IconPositionProps;

const inputContentModifiers = {
  withIcon: (theme: DefaultTheme, iconPosition: "left" | "right") => css`
    border-${iconPosition}: 0.1rem solid ${theme.colors.mainBg};
    padding-${iconPosition}: ${theme.spacings.xsmall};
    margin-${iconPosition}: ${theme.spacings.xsmall};

    ${Label} {
      left: ${iconPosition === "left" ? theme.spacings.xsmall : 0};
    }
  `
};

export const InputContent = styled.div<InputContentProps>`
  ${({ theme, hasIcon, iconPosition }) => css`
    flex: 1;
    position: relative;

    ${hasIcon && inputContentModifiers.withIcon(theme, iconPosition)}
  `}
`;

type InputProps = {
  hasLabel: boolean;
} & HasValueProps;

export const Input = styled.input<InputProps>`
  ${({ theme, hasValue, hasLabel }) => css`
    font-size: ${theme.font.sizes.small};
    font-weight: ${theme.font.light};
    color: ${theme.colors.gray700};
    background: transparent;
    width: 100%;
    padding: ${theme.spacings.xsmall} 0;
    border: 0;
    outline: none;
    transition: 0.4s;
    transform: ${hasValue && hasLabel ? "translateY(0.8rem)" : "translateY(0)"};

    ${hasLabel &&
    css`
      &:focus {
        transform: translateY(0.8rem);
        transition: 0.4s;

        & + label {
          opacity: 1;
          transition: 0.4s;
        }
      }
    `}

    &::placeholder {
      font-size: ${theme.font.sizes.small};
      font-weight: ${theme.font.light};
      color: ${theme.colors.gray400};
    }
  `}
`;

export const Label = styled.label<HasValueProps>`
  ${({ theme, hasValue }) => css`
    cursor: text;
    position: absolute;
    background: transparent;
    color: ${theme.colors.gray400};
    font-size: ${theme.font.sizes.xxsmall};
    font-weight: ${theme.font.light};
    top: 0.6rem;
    left: 0;
    opacity: ${hasValue ? "1" : "0"};
    transition: 0.4s;
  `}
`;

export const Icon = styled.div<IconPositionProps>`
  ${({ theme, iconPosition }) => css`
    display: flex;
    align-items: center;
    justify-content: center;
    order: ${iconPosition === "right" ? 1 : 0};

    svg {
      width: 2rem;
      height: 2rem;
      color: ${theme.colors.gray500};
    }
  `}
`;

export const Error = styled.p`
  ${({ theme }) => css`
    color: ${theme.colors.lightPrimary};
    font-size: ${theme.font.sizes.xsmall};
  `}
`;

const wrapperModifiers = {
  disabled: () => css`
    ${Label},
    ${Input},
    ${InputWrapper} {
      cursor: not-allowed;
    }

    ${InputWrapper} {
      opacity: 0.8;
    }
  `,

  error: (theme: DefaultTheme) => css`
    ${Label} {
      color: ${theme.colors.lightPrimary};
    }

    ${InputWrapper} {
      border: 0.1rem solid ${theme.colors.lightPrimary};
    }
  `,

  withValue: (theme: DefaultTheme) => css`
    ${Icon} {
      svg {
        color: ${theme.colors.primary};
      }
    }
  `
};

export const Wrapper = styled.div<WrapperProps>`
  ${({ theme, disabled, error, hasValue }) => css`
    ${disabled && wrapperModifiers.disabled()}
    ${error && wrapperModifiers.error(theme)}
    ${hasValue && wrapperModifiers.withValue(theme)}
  `}
`;
