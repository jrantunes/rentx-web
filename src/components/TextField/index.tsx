import { ChangeEvent, InputHTMLAttributes, ReactNode, useState } from "react";

import * as S from "./styles";

export type TextFieldProps = {
  onInputChange?: (value: string) => void;
  icon?: ReactNode;
  iconPosition?: "left" | "right";
  label?: string;
  labelFor?: string;
  initialValue?: string;
  disabled?: boolean;
  error?: string;
} & InputHTMLAttributes<HTMLInputElement>;

const TextField = ({
  onInputChange,
  icon,
  label,
  error,
  labelFor = "",
  initialValue = "",
  iconPosition = "left",
  disabled = false,
  ...props
}: TextFieldProps) => {
  const [value, setValue] = useState(initialValue);

  const onChange = (e: ChangeEvent<HTMLInputElement>) => {
    const newValue = e.currentTarget.value;

    setValue(newValue);

    !!onInputChange && onInputChange(newValue);
  };

  return (
    <S.Wrapper disabled={disabled} error={!!error} hasValue={!!value}>
      <S.InputWrapper>
        {!!icon && <S.Icon iconPosition={iconPosition}>{icon}</S.Icon>}
        <S.InputContent iconPosition={iconPosition} hasIcon={!!icon}>
          <S.Input
            id={labelFor}
            value={value}
            disabled={disabled}
            hasValue={!!value}
            hasLabel={!!label}
            onChange={onChange}
            autoComplete="off"
            {...props}
          />
          {!!label && (
            <S.Label aria-hidden={!value} hasValue={!!value} htmlFor={labelFor}>
              {label}
            </S.Label>
          )}
        </S.InputContent>
      </S.InputWrapper>
      {!!error && <S.Error>{error}</S.Error>}
    </S.Wrapper>
  );
};

export default TextField;
