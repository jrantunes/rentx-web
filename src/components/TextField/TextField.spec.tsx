import { screen, waitFor } from "@testing-library/react";
import { renderWithTheme } from "utils/tests/helpers";
import userEvent from "@testing-library/user-event";

import TextField from ".";
import { AiOutlineMail } from "react-icons/ai";

describe("<TextField />", () => {
  it("should render with label", () => {
    renderWithTheme(
      <TextField label="Label" labelFor="input" placeholder="Placeholder" />
    );

    expect(screen.getByLabelText(/label/i)).toBeInTheDocument();
  });

  it("should render without label", () => {
    renderWithTheme(<TextField placeholder="Placeholder" />);

    expect(screen.queryByLabelText(/label/i)).not.toBeInTheDocument();
  });

  it("should render with placeholder", () => {
    renderWithTheme(
      <TextField label="Label" labelFor="input" placeholder="Placeholder" />
    );

    expect(screen.getByPlaceholderText(/placeholder/i)).toBeInTheDocument();
  });

  it("should change its value when typing", async () => {
    const onInputChange = jest.fn();

    renderWithTheme(
      <TextField
        label="Label"
        labelFor="input"
        placeholder="Placeholder"
        onInputChange={onInputChange}
      />
    );

    const input = screen.getByPlaceholderText(/placeholder/i);
    const text = "Text example";

    userEvent.type(input, text);

    await waitFor(() => {
      expect(onInputChange).toHaveBeenCalledTimes(text.length);
      expect(input).toHaveValue(text);
    });

    expect(onInputChange).toHaveBeenCalledWith(text);
  });

  it("should not change its value when disabled", async () => {
    const onInputChange = jest.fn();

    renderWithTheme(
      <TextField
        label="Label"
        labelFor="input"
        placeholder="Placeholder"
        onInputChange={onInputChange}
        disabled
      />
    );

    const input = screen.getByPlaceholderText(/placeholder/i);
    expect(input).toBeDisabled();

    const text = "Text example";

    userEvent.type(input, text);

    await waitFor(() => {
      expect(input).not.toHaveValue(text);
    });

    expect(onInputChange).not.toHaveBeenCalled();
  });

  it("should render label with dynamic opacity based on input value", async () => {
    renderWithTheme(
      <TextField label="Label" labelFor="input" placeholder="Placeholder" />
    );

    const input = screen.getByPlaceholderText(/placeholder/i);
    const label = screen.getByText(/label/i);

    expect(label).toHaveStyle({
      opacity: "0"
    });

    expect(label).toHaveAttribute("aria-hidden", "true");

    userEvent.type(input, "Some test");

    await waitFor(() => {
      expect(input).toHaveFocus();
      expect(label).toHaveStyle({
        opacity: "1"
      });
    });

    expect(label).toHaveAttribute("aria-hidden", "false");
  });

  it("should be accessible with tab", async () => {
    renderWithTheme(
      <TextField label="Label" labelFor="input" placeholder="Placeholder" />
    );

    const input = screen.getByPlaceholderText(/placeholder/i);

    expect(document.body).toHaveFocus();

    userEvent.tab();

    await waitFor(() => {
      expect(document.body).not.toHaveFocus();
      expect(input).toHaveFocus();
    });
  });

  it("should not to be accessible with tab when disabled", async () => {
    renderWithTheme(
      <TextField
        label="Label"
        labelFor="input"
        placeholder="Placeholder"
        disabled
      />
    );

    const input = screen.getByPlaceholderText(/placeholder/i);

    expect(document.body).toHaveFocus();

    userEvent.tab();

    await waitFor(() => {
      expect(input).not.toHaveFocus();
    });
  });

  it("should render with error", () => {
    renderWithTheme(
      <TextField
        label="Label"
        labelFor="input"
        placeholder="Placeholder"
        error="Error Message"
      />
    );

    expect(screen.getByText(/error message/i)).toBeInTheDocument();
  });

  it("should render with icon", () => {
    const { container } = renderWithTheme(
      <TextField
        label="Label"
        labelFor="input"
        placeholder="Placeholder"
        error="Error Message"
        icon={<AiOutlineMail data-testid="icon" />}
      />
    );

    expect(screen.getByTestId("icon")).toBeInTheDocument();

    expect(container.firstChild).toMatchSnapshot();
  });

  it("should render icon on the left side by default", () => {
    renderWithTheme(
      <TextField
        label="Label"
        labelFor="input"
        placeholder="Placeholder"
        error="Error Message"
        icon={<AiOutlineMail data-testid="icon" />}
      />
    );

    expect(screen.getByTestId("icon").parentElement).toHaveStyle({
      order: 0
    });
  });

  it("should render icon on the right side", () => {
    renderWithTheme(
      <TextField
        label="Label"
        labelFor="input"
        placeholder="Placeholder"
        error="Error Message"
        iconPosition="right"
        icon={<AiOutlineMail data-testid="icon" />}
      />
    );

    expect(screen.getByTestId("icon").parentElement).toHaveStyle({
      order: 1
    });
  });
});
