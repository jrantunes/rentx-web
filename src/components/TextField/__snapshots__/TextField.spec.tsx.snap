// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TextField /> should render with icon 1`] = `
.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background-color: #FFFFFF;
  padding: 0 1.6rem;
  border: 0.1rem solid #EBEBF0;
  -webkit-transition: border 0.1s ease-in-out;
  transition: border 0.1s ease-in-out;
  cursor: text;
}

.c2:hover,
.c2:focus-within {
  border-color: #AEAEB3;
}

.c4 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  position: relative;
  border-left: 0.1rem solid #F4F5F6;
  padding-left: 1.6rem;
  margin-left: 1.6rem;
}

.c4 .c6 {
  left: 1.6rem;
}

.c5 {
  font-size: 1.6rem;
  font-weight: 400;
  color: #47474D;
  background: transparent;
  width: 100%;
  padding: 1.6rem 0;
  border: 0;
  outline: none;
  -webkit-transition: 0.4s;
  transition: 0.4s;
  -webkit-transform: translateY(0);
  -ms-transform: translateY(0);
  transform: translateY(0);
}

.c5:focus {
  -webkit-transform: translateY(0.8rem);
  -ms-transform: translateY(0.8rem);
  transform: translateY(0.8rem);
  -webkit-transition: 0.4s;
  transition: 0.4s;
}

.c5:focus + label {
  opacity: 1;
  -webkit-transition: 0.4s;
  transition: 0.4s;
}

.c5::-webkit-input-placeholder {
  font-size: 1.6rem;
  font-weight: 400;
  color: #AEAEB3;
}

.c5::-moz-placeholder {
  font-size: 1.6rem;
  font-weight: 400;
  color: #AEAEB3;
}

.c5:-ms-input-placeholder {
  font-size: 1.6rem;
  font-weight: 400;
  color: #AEAEB3;
}

.c5::placeholder {
  font-size: 1.6rem;
  font-weight: 400;
  color: #AEAEB3;
}

.c7 {
  cursor: text;
  position: absolute;
  background: transparent;
  color: #AEAEB3;
  font-size: 1.2rem;
  font-weight: 400;
  top: 0.6rem;
  left: 0;
  opacity: 0;
  -webkit-transition: 0.4s;
  transition: 0.4s;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-order: 0;
  -ms-flex-order: 0;
  order: 0;
}

.c3 svg {
  width: 2rem;
  height: 2rem;
  color: #7A7A80;
}

.c8 {
  color: #FF99AA;
  font-size: 1.4rem;
}

.c0 .c6 {
  color: #FF99AA;
}

.c0 .c1 {
  border: 0.1rem solid #FF99AA;
}

<div
  class="c0"
>
  <div
    class="c1 c2"
  >
    <div
      class="c3"
    >
      <svg
        data-testid="icon"
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 1024 1024"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M928 160H96c-17.7 0-32 14.3-32 32v640c0 17.7 14.3 32 32 32h832c17.7 0 32-14.3 32-32V192c0-17.7-14.3-32-32-32zm-40 110.8V792H136V270.8l-27.6-21.5 39.3-50.5 42.8 33.3h643.1l42.8-33.3 39.3 50.5-27.7 21.5zM833.6 232L512 482 190.4 232l-42.8-33.3-39.3 50.5 27.6 21.5 341.6 265.6a55.99 55.99 0 0 0 68.7 0L888 270.8l27.6-21.5-39.3-50.5-42.7 33.2z"
        />
      </svg>
    </div>
    <div
      class="c4"
    >
      <input
        autocomplete="off"
        class="c5"
        id="input"
        placeholder="Placeholder"
        value=""
      />
      <label
        aria-hidden="true"
        class="c6 c7"
        for="input"
      >
        Label
      </label>
    </div>
  </div>
  <p
    class="c8"
  >
    Error Message
  </p>
</div>
`;
